source/Calamari.Aws/Deployment/Conventions/BucketKeyProvider.cs

```diff
@@ -1,26 +1,93 @@
+            if (!packageFilePath.IsNullOrEmpty())
+            {
+                packageContentHash = CalculateContentHash(packageFilePath);
+            }
             switch (behaviour.BucketKeyBehaviour)
             {
                 case BucketKeyBehaviourType.Custom:
-                    return behaviour.BucketKey;
+                    return SubstitutePackageHashVariable(behaviour.BucketKey, packageContentHash);
                 case BucketKeyBehaviourType.Filename:
-                    return $"{behaviour.BucketKeyPrefix}{defaultKey}";
+                    return $"{SubstitutePackageHashVariable(behaviour.BucketKeyPrefix, packageContentHash)}{defaultKey}";
+                case BucketKeyBehaviourType.FilenameWithContentHash:
+                    var (fileName, extension) = GetFileNameParts(defaultKey);
+                    var bucketKey = new StringBuilder();
+                    bucketKey.Append(SubstitutePackageHashVariable(behaviour.BucketKeyPrefix, packageContentHash));
+                    bucketKey.Append(fileName);
+                    if (!packageContentHash.IsNullOrEmpty()) bucketKey.Append($"@{packageContentHash}");
+                    bucketKey.Append(extension);
+                    return bucketKey.ToString();
                 default:
                     throw new ArgumentOutOfRangeException(nameof(behaviour), "The provided bucket key behavior was not valid.");
             }
         }
+
+        string CalculateContentHash(string packageFilePath)
+        {
+            var packageContent = File.ReadAllBytes(packageFilePath);
+            using (SHA256 sha256Hash = SHA256.Create())
+            {
+                var computedHashByte = sha256Hash.ComputeHash(packageContent);
+                var computedHash = new StringBuilder();
+                foreach (var c in computedHashByte)
+                {
+                    computedHash.Append(c.ToString("X2"));
+                }
+
+                return computedHash.ToString();
+            }
+        }
+
+        string SubstitutePackageHashVariable(string input, string contentHash)
+        {
+            return input.Replace($"#{{Octopus.Action.Package.PackageContentHash}}", contentHash);
+        }
+        
+        public string EncodeBucketKeyForUrl(string bucketKey)
+        {
+            var prefix = Path.GetDirectoryName(bucketKey)?.Replace('\\','/') ?? string.Empty;
+            var (fileName, extension) = GetFileNameParts(Path.GetFileName(bucketKey));
+            return $"{prefix}/{Uri.EscapeDataString(fileName)}{extension}";
+        }
+
+        (string filename, string extension) GetFileNameParts(string fileNameWithExtensions)
+        {
+            var fileNameParts = fileNameWithExtensions.Split('.');
+            var fileName = fileNameParts[0];
+            var extension = string.Empty;
+            for (var i = 1; i < fileNameParts.Length; i++)
+            {
+                if (int.TryParse(fileNameParts[i], out _)) 
```

Comment:
I am not sure if this is correct. Consider filename `helloworld.1.0.0-beta.zip` where version `1.0.0` is not an integer. Perhaps you could first use a built-in method to get the filename (with version) and extension? And we have probably implemented extracting the version somewhere else already.

Reply:
thanks for picking this up. I added a test `GetBucketKey_PackageOptions_ShouldAppendContentHash_WhenFileNameHasVersionNumbersAndReleaseTag` to cover this case.

I tried to use the PackageName parser but for some reason, it doesn't work well with BucketKey. I looked at [the list of extensions that we supported](https://octopus.com/docs/packaging-applications) and see the only multiple-parts extensions are `.tar.something`. Thus, I changed my implementation a bit to handled only `.tar.something` if the extensions may have multiple parts.